modules/default_instance/main.tf

```diff
@@ -0,0 +1,119 @@
+}
+
+/******************************************
+  Firewall rule to allow tunnel-through-iap
+ *****************************************/
+module "iap_firewall" {
+  source               = "../iap_firewall"
+  project_id           = var.project_id
+  network_name         = var.network_name
+  firewall_description = "Allow SSH and web UI IAP tunnel to Datalab instance"
+  target_tags          = ["datalab"]
+  ports                = ["22", "8080"]
+}
+
+/***********************************************
+  Render start up and cloud config scripts from templates
+ ***********************************************/
+module "template_files" {
+  source                    = "../template_files"
+  cloud_config              = var.cloud_config
+  datalab_disk_name         = local.datalab_disk_name
+  datalab_enable_swap       = var.datalab_enable_swap
+  datalab_docker_image      = var.datalab_docker_image
+  datalab_enable_backup     = var.datalab_enable_backup
+  datalab_console_log_level = var.datalab_console_log_level
+  datalab_user_email        = var.datalab_user_email
+  datalab_idle_timeout      = var.datalab_idle_timeout
+  fluentd_docker_image      = var.fluentd_docker_image
+}
+
+/***********************************************
+  Create GCE instance
+ ***********************************************/
+resource "google_compute_instance" "main" {
+  name                      = var.name
+  project                   = var.project_id
+  machine_type              = var.machine_type
+  zone                      = var.zone
+  allow_stopping_for_update = true
+
+  tags = ["datalab"]
+
+  labels = {
+    role    = "datalab"
+    use_gpu = false
+  }
+
+  boot_disk {
+    initialize_params {
+      image = "cos-cloud/cos-stable"
+      size  = var.boot_disk_size_gb
+    }
+  }
+
+  attached_disk {
+    source      = var.existing_disk_name == null ? google_compute_disk.main.0.name : var.existing_disk_name
+    device_name = local.datalab_disk_name
+  }
+
+  network_interface {
+    subnetwork = var.subnet_name
+  }
+
+  metadata = {
+    enable-oslogin = "TRUE"
+    for-user       = var.datalab_user_email
+    user-data      = module.template_files.cloud_config
+  }
+
+  metadata_startup_script = module.template_files.startup_script
+
+  service_account {
+    email  = var.service_account
+    scopes = ["cloud-platform"]
+  }
+
+  shielded_instance_config {
+    enable_integrity_monitoring = true
+    enable_secure_boot          = false
+    enable_vtpm                 = true
+  }
+
+  scheduling {
+    automatic_restart   = true
+    on_host_maintenance = "MIGRATE"
+    preemptible         = false
+  }
+}
+
+/******************************************
+  Create the persistent data disk
+  Does not create a new disk if an existing_disk_name is set
+ *****************************************/
+resource "google_compute_disk" "main" {
+  name    = local.datalab_disk_name
+  project = var.project_id
+  count   = var.existing_disk_name == null ? 1 : 0
```

Comment:
This logic can be problematic if `existing_disk_name` is populated with a dynamic value. We should use a separate `create_disk` boolean variable to drive this.

Reply:
Added separate create_disk boolean variable